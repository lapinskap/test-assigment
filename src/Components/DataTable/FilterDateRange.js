import PropTypes from 'prop-types';
import React from 'react';

import {
  InputGroup,
  InputGroupAddon,
  FormGroup,
  Row,
} from 'reactstrap';

import { faCalendarAlt } from '@fortawesome/free-solid-svg-icons';

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import DatePicker from 'react-datepicker';
import hashInputId from '../../utils/jsHelpers/hashInputId';

export default function FilterDateRange({
  value, onChange, id, showTimeSelect,
}) {
  const inputValue = value;

  const handleChange = ({ from, to }) => {
    const fromValue = from === undefined ? inputValue.from : from;
    let toValue = to === undefined ? inputValue.to : to;

    if (toValue && fromValue > toValue) {
      toValue = fromValue;
    }
    onChange({
      from: fromValue,
      to: toValue,
    });
  };

  const handleChangeStart = (from) => handleChange({ from });
  const handleChangeEnd = (to) => {
    let newValue = to;
    if (!showTimeSelect && newValue) {
      newValue = new Date(newValue);
      newValue.setHours(23);
      newValue.setMinutes(59);
      newValue.setSeconds(59);
    }
    return handleChange({ to: newValue });
  };

  const from = value && value.from ? new Date(value.from) : null;
  const to = value && value.to ? new Date(value.to) : null;

  return (
    <>
      <Row form>
        <div style={{ width: '47.5%' }}>
          <FormGroup>
            <InputGroup>
              <InputGroupAddon addonType="prepend">
                <div className="input-group-text">
                  <FontAwesomeIcon icon={faCalendarAlt} />
                </div>
              </InputGroupAddon>
              <div
                className={`m-0 p-0 ${showTimeSelect ? 'datetime-picker-wrapper' : 'date-picker-wrapper'}`}
                data-t1="gridFilter"
                data-t2={`${id}From`}
              >
                <DatePicker
                  id={`${hashInputId(id)}_from`}
                  popperPlacement="auto"
                  popperClassName="some-custom-class"
                  popperModifiers={{
                    preventOverflow: {
                      enabled: true,
                    },
                  }}
                  selected={from}
                  placeholderText="od"
                  selectsStart
                  showTimeSelect={showTimeSelect}
                  className="form-control"
                  startDate={from}
                  endDate={to}
                  onChange={handleChangeStart}
                  dateFormat={showTimeSelect ? 'dd/MM/yyyy H:mm' : 'dd/MM/yyyy'}
                />
              </div>
            </InputGroup>
          </FormGroup>
        </div>
        <p style={{ width: '5%' }}>
          -
        </p>
        <div style={{ width: '47.5%' }}>
          <FormGroup>
            <InputGroup>
              <InputGroupAddon addonType="prepend">
                <div className="input-group-text">
                  <FontAwesomeIcon icon={faCalendarAlt} />
                </div>
              </InputGroupAddon>
              <div
                className={`m-0 p-0 ${showTimeSelect ? 'datetime-picker-wrapper' : 'date-picker-wrapper'}`}
                data-t1="gridFilter"
                data-t2={`${id}To`}
              >
                <DatePicker
                  id={`${hashInputId(id)}_to`}
                  popperPlacement="auto"
                  popperModifiers={{
                    preventOverflow: {
                      enabled: true,
                    },
                  }}
                  selected={to}
                  showTimeSelect={showTimeSelect}
                  placeholderText="do"
                  selectsEnd
                  className="form-control"
                  startDate={from}
                  endDate={to}
                  onChange={handleChangeEnd}
                  timeFormat="HH:mm"
                  dateFormat={showTimeSelect ? 'dd/MM/yyyy HH:mm aa' : 'dd/MM/yyyy'}
                />
              </div>
            </InputGroup>
          </FormGroup>
        </div>
      </Row>
    </>
  );
}

FilterDateRange.propTypes = {
  id: PropTypes.string,
  onChange: PropTypes.func,
  showTimeSelect: PropTypes.bool.isRequired,
  value: PropTypes.shape({
    from: PropTypes.oneOfType([PropTypes.instanceOf(Date), PropTypes.string]),
    to: PropTypes.oneOfType([PropTypes.instanceOf(Date), PropTypes.string]),
  }),
};

FilterDateRange.defaultProps = {
  id: '',
  onChange: null,
  value: {},
};
