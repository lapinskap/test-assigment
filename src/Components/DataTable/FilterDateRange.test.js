import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import FilterDateRange from './FilterDateRange';

describe('<FilterDateRange />', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  global.document.createRange = () => ({
    setStart: () => {},
    setEnd: () => {},
    commonAncestorContainer: {
      nodeName: 'BODY',
      ownerDocument: document,
    },
  });

  it('renders correctly with props', () => {
    const format = 'dd-MM-yyy';
    const wrapper = renderer
      .create(<FilterDateRange format={format} id="test" />);

    expect(wrapper.toJSON())
      .toMatchSnapshot();
  });

  test('FROM input acts correctly on change when time select is off', () => {
    const onChangeSpy = jest.fn(({ from }) => {
      expect(from.toString()).toBe(new Date('01/01/2021').toString());
    });
    const format = 'dd-MM-yyy';

    const { container } = render(<FilterDateRange format={format} id="test" onChange={onChangeSpy} showTimeSelect={false} />);
    const input = container.querySelector('#test_from');

    fireEvent.change(input, { target: { value: '01/01/2021' } });
    expect(onChangeSpy).toHaveBeenCalledTimes(1);
  });

  test('TO input acts correctly on change when time select is off', () => {
    const onChangeSpy = jest.fn(({ to }) => {
      expect(+to).toBe(+new Date('01/01/2021') + (24 * 60 * 60 * 1000) - 1000);
    });
    const format = 'dd-MM-yyy';

    const { container } = render(<FilterDateRange format={format} id="test" onChange={onChangeSpy} showTimeSelect={false} />);
    const input = container.querySelector('#test_to');

    fireEvent.change(input, { target: { value: '01/01/2021' } });
    expect(onChangeSpy).toHaveBeenCalledTimes(1);
  });

  test('TO input acts correctly on change when time select is on', () => {
    const onChangeSpy = jest.fn(({ to }) => {
      expect(to.toString()).toBe(new Date('01/01/2021').toString());
    });
    const format = 'dd-MM-yyy';

    const { container } = render(<FilterDateRange format={format} id="test" onChange={onChangeSpy} showTimeSelect />);
    const input = container.querySelector('#test_to');

    fireEvent.change(input, { target: { value: '01/01/2021' } });
    expect(onChangeSpy).toHaveBeenCalledTimes(1);
  });

  test('TO cannot be earlier than from', () => {
    const fromValue = new Date('2021-01-01');
    const onChangeSpy = jest.fn(({ to }) => {
      expect(to.toString()).toBe(fromValue.toString());
    });
    const format = 'dd-MM-yyy';

    const { container } = render(<FilterDateRange format={format} id="test" onChange={onChangeSpy} showTimeSelect value={{ from: fromValue }} />);
    const input = container.querySelector('#test_to');

    fireEvent.change(input, { target: { value: '12/31/2020' } });
    expect(onChangeSpy).toHaveBeenCalledTimes(1);
  });
});
