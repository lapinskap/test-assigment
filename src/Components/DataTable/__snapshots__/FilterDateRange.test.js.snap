// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilterDateRange /> renders correctly with props 1`] = `
<div
  className="form-row"
>
  <div
    style={
      Object {
        "width": "47.5%",
      }
    }
  >
    <div
      className="form-group"
    >
      <div
        className="input-group"
      >
        <div
          className="input-group-prepend"
        >
          <div
            className="input-group-text"
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-calendar-alt fa-w-14 "
              data-icon="calendar-alt"
              data-prefix="fas"
              focusable="false"
              role="img"
              style={Object {}}
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V192H0v272zm320-196c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12v-40zm0 128c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12v-40zM192 268c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12v-40zm0 128c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12v-40zM64 268c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12H76c-6.6 0-12-5.4-12-12v-40zm0 128c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12H76c-6.6 0-12-5.4-12-12v-40zM400 64h-48V16c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v48H160V16c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v48H48C21.5 64 0 85.5 0 112v48h448v-48c0-26.5-21.5-48-48-48z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </div>
        </div>
        <div
          className="m-0 p-0 date-picker-wrapper"
          data-t1="gridFilter"
          data-t2="testFrom"
        >
          <div
            className="react-datepicker-wrapper"
          >
            <div
              className="react-datepicker__input-container"
            >
              <input
                className="form-control"
                disabled={false}
                id="test_from"
                onBlur={[Function]}
                onChange={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                placeholder="od"
                readOnly={false}
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <p
    style={
      Object {
        "width": "5%",
      }
    }
  >
    -
  </p>
  <div
    style={
      Object {
        "width": "47.5%",
      }
    }
  >
    <div
      className="form-group"
    >
      <div
        className="input-group"
      >
        <div
          className="input-group-prepend"
        >
          <div
            className="input-group-text"
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-calendar-alt fa-w-14 "
              data-icon="calendar-alt"
              data-prefix="fas"
              focusable="false"
              role="img"
              style={Object {}}
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V192H0v272zm320-196c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12v-40zm0 128c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12v-40zM192 268c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12v-40zm0 128c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12v-40zM64 268c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12H76c-6.6 0-12-5.4-12-12v-40zm0 128c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12H76c-6.6 0-12-5.4-12-12v-40zM400 64h-48V16c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v48H160V16c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v48H48C21.5 64 0 85.5 0 112v48h448v-48c0-26.5-21.5-48-48-48z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </div>
        </div>
        <div
          className="m-0 p-0 date-picker-wrapper"
          data-t1="gridFilter"
          data-t2="testTo"
        >
          <div
            className="react-datepicker-wrapper"
          >
            <div
              className="react-datepicker__input-container"
            >
              <input
                className="form-control"
                disabled={false}
                id="test_to"
                onBlur={[Function]}
                onChange={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                placeholder="do"
                readOnly={false}
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
